stream.go

```diff
@@ -96,11 +101,30 @@ func (s *stream) Read(p []byte) (int, error) {
 				err = s.err
 				break
 			}
+
+			s.mutex.Lock()
+			deadline := s.readDeadline
+			s.mutex.Unlock()
+
+			if !deadline.IsZero() && !time.Now().Before(deadline) {
+				err = errDeadline
+				break
+			}
+
 			if frame != nil {
 				s.readPosInFrame = int(s.readOffset - frame.Offset)
 				break
 			}
-			<-s.readChan
+
+			if deadline.IsZero() {
+				<-s.readChan
+			} else {
+				select {
+				case <-s.readChan:
+				case <-time.After(deadline.Sub(time.Now())):
```

Comment:
There's a race here: You should only call time.Now() once, and use the value for the comparison above and the duration calculation here. Otherwise, the condition above might be false, but here the duration is negative (which leads to the timer now firing).